metricflow/time/time_granularity_solver.py

```diff
@@ -156,14 +156,16 @@ def validate_time_granularity(
                 # If there is a cumulative metric, granularity changes aren't supported.
                 for metric_reference in metric_references:
                     metric = self._semantic_model.metric_semantics.get_metric(metric_reference)
-                    if (
-                        metric.type == MetricType.CUMULATIVE
-                        and time_dimension_spec.time_granularity != min_granularity_for_querying
-                    ):
-                        raise RequestTimeGranularityException(
-                            f"For querying cumulative metric '{metric_reference.element_name}', the granularity of "
-                            f"'{time_dimension_spec.qualified_name}' must be {min_granularity_for_querying.name}"
+                    if metric.type == MetricType.CUMULATIVE:
+                        _, only_queryable_granularity = self.local_dimension_granularity_range(
+                            metric_references=[metric_reference],
+                            local_time_dimension_reference=time_dimension_spec.reference,
```

Comment:
Ooof, all of our granularity resolution stuff is impossibly complicated. :(

I don't know if this is works correctly in all cases. The time_dimension_spec could be pulled in from a join, in which case it's not clear to me what happens. To be honest, it's not clear to me what happens if we somehow have multiple "local" time_dimension_specs, either, and of course that's been sitting in this code path since we open sourced.

I think all of this logic was implicitly tied to the old model where everybody queried by a single primary time dimension which was consistent model-wide (essentially, `metric_time`). But I'd expect that to work correctly already - like if `metric_time` was in the wrong granularity we should throw an error here.

Reply:
There is some weirdness about how we get here for sure - for example, if I query in MetricFlow  `mf --metrics wau_rolling_7,gusto_survey_response_rate --dimensions metric_time__month`, I get the expected error `For querying cumulative metric 'wau_rolling_7', the granularity of 'metric_time__month' must be DAY`.
But if I run the pseudo-equivalent query in Transform `mql query --metrics wau_rolling_7,gusto_survey_response_rate --dimensions metric_time__month` I unexpectedly get a successful result. Working on figuring out why that difference occurs. (But this change fixes it on the Transform side.)

But the logic itself here seems flawed regardless - `local_dimension_granularity_range`  would return `[DAY, MONTH]` for the 2 metrics in question. We then treat `MONTH` as the minimum queryable granularity for BOTH metrics, even though `wau_rolling_7` only supports `DAY`. For cumulative metrics, we need to check their supported granularity individually.

Reply:
Ok, I get it now. The solver is confusing because it doesn't really understand metric types. In the long run we'll need to address this, but for now I think this is the right fix. Let's do one of the following just to make it more obvious why this is here:

1. Add a comment that we have to do this again for cumulative metrics because the range method might return the wrong value
2. Factor the granularity range extraction for a single metric into its own method (instead of having it all inlined in `local_dimension_granularity_range`) and then assert against that here.